src/packet_analysis/Manager.cc

```diff
@@ -171,3 +182,69 @@ void Manager::DumpPacket(const Packet *pkt, int len)
 
 	run_state::detail::pkt_dumper->Dump(pkt);
 	}
+
+class UnknownProtocolTimer final : public zeek::detail::Timer {
+public:
+	// Represents a combination of an analyzer name and protocol identifier, where the identifier was
+	// reported as unknown by the analyzer.
+	using UnknownProtocolPair = std::pair<std::string, uint32_t>;
+
+	UnknownProtocolTimer(double t, UnknownProtocolPair p, double timeout)
+		: zeek::detail::Timer(t + timeout, zeek::detail::TIMER_UNKNOWN_PROTOCOL_EXPIRE),
+		  unknown_protocol(std::move(p))
+		{}
+
+	void Dispatch(double t, bool is_expire) override
+		{ zeek::packet_mgr->ResetUnknownProtocolTimer(unknown_protocol.first, unknown_protocol.second); }
+
+	UnknownProtocolPair unknown_protocol;
+};
+
+void Manager::ResetUnknownProtocolTimer(const std::string& analyzer, uint32_t protocol)
+	{
+	unknown_protocols.erase(std::make_pair(analyzer, protocol));
+	}
+
+bool Manager::PermitUnknownProtocol(const std::string& analyzer, uint32_t protocol)
+	{
+	auto p = std::make_pair(analyzer, protocol);
+	uint64_t& count = unknown_protocols[p];
+	++count;
+
+	if ( count == 1 )
+		detail::timer_mgr->Add(new UnknownProtocolTimer(run_state::network_time, p,
+		                                                unknown_sampling_duration));
+
+	if ( count < unknown_sampling_threshold )
+		return true;
+
+	auto num_above_threshold = count - unknown_sampling_threshold;
+	if ( unknown_sampling_rate )
+		return num_above_threshold % unknown_sampling_rate == 0;
+
+	return false;
+	}
+
+void Manager::ReportUnknownProtocol(const std::string& analyzer, uint32_t protocol,
+                                    const uint8_t* data, size_t len)
+	{
+	if ( unknown_protocol )
+		{
+		if ( PermitUnknownProtocol(analyzer, protocol ) )
+			{
+			std::stringstream ss;
+
+			if ( data )
+				{
+				ss << std::hex;
+				for ( int i = 0; i < unknown_first_bytes_count && i < len; i++ )
+					ss << std::setw(2) << std::setfill('0') << (int)data[i];
+				}
+
+			event_mgr.Enqueue(unknown_protocol,
+			                  make_intrusive<StringVal>(analyzer),
+			                  val_mgr->Count(protocol),
+			                  make_intrusive<StringVal>(ss.str()));
```

Comment:
Seems like this this could use `StringVal(int, const char*)` constructor directly from the `data` buffer instead of going through a `stringstream` first.

Reply:
It goes through the `stringstream` so it can be printed as hex values.

Reply:
Is it important to pre-format ?  Most times I see payload data passed to scripts via events it's the raw bytes stuffed into a `string` parameter and left up to consumer how/when to format (e.g. with `bytestring_to_hexstr()`), so it's throwing my expectations off a bit (alternatively, call the pre-formatting out in the `unknown_protocol` event docs since I overlooked it).

Reply:
Ah ha! That's the function that neither @sethhall or I could think of. I can change it over to that.